Assets/MixedRealityToolkit/Utilities/BuildAndDeploy/UwpAppxBuildTools.cs

```diff
@@ -408,6 +438,125 @@ public static void AddGazeInputCapability(IBuildInfo buildInfo)
+                ignNsAttribute.Value += " rescap";
+            }
+
+            // If the capabilities container tag is missing, make sure it gets added.
+            var capabilitiesTag = rootNode.GetDefaultNamespace() + "Capabilities";
+            XElement capabilitiesNode = rootNode.Element(capabilitiesTag);
+            if (capabilitiesNode == null)
+            {
+                capabilitiesNode = new XElement(capabilitiesTag);
+                rootNode.Add(capabilitiesNode);
+            }
+
+            var researchModeCapability = rescapNs + "Capability";
+            XElement existingResearchModeCapability = capabilitiesNode.Elements(researchModeCapability)
+                .FirstOrDefault(element => element.Attribute("Name")?.Value == "perceptionSensorsExperimental");
+
+            // Only add the capability if isn't there already.
+            if (existingResearchModeCapability == null)
+            {
+                capabilitiesNode.Add(
+                    new XElement(researchModeCapability, new XAttribute("Name", "perceptionSensorsExperimental")));
+            }
+        }
+
+        /// <summary>
+        /// An overload of AddResearchModeCapability that will read the AppX manifest
+        /// from the build output and update the manifest file with the rescap aka
+        /// research mode capability.
+        /// </summary>
+        /// <param name="buildInfo">An IBuildInfo containing a valid OutputDirectory</param>
+        public static void AddResearchModeCapability(IBuildInfo buildInfo)
+        {
+            string manifestFilePath = GetManifestFilePath(buildInfo);
+            if (manifestFilePath == null)
+            {
+                throw new FileNotFoundException("Unable to find manifest file");
+            }
+
+            var rootElement = XElement.Load(manifestFilePath);
+            AddResearchModeCapability(rootElement);
+            rootElement.Save(manifestFilePath);
+        }
+
+        /// <summary>
+        /// Enables unsafe code in the generated Assembly-CSharp project.
+        /// </summary>
+        /// <remarks>
+        /// This is not required by the research mode, but not using unsafe code with
+        /// direct memory access results in poor performance. So its kinda recommended
+        /// to use unsafe code.
+        /// For further information take a look at https://docs.microsoft.com/en-us/windows/mixed-reality/research-mode.
+        /// Note that this function is only public to poke a hole for testing - do not
+        /// take a dependency on this function.
+        /// </remarks>
+        public static void AllowUnsafeCode(XElement rootNode)
+        {
+            foreach (XElement propertyGroupNode in rootNode.Descendants(rootNode.GetDefaultNamespace() + "PropertyGroup"))
+            {
+                if (propertyGroupNode.Attribute("Condition") != null)
+                {
+                    var allowUnsafeBlock = propertyGroupNode.Element("AllowUnsafeBlock");
```

Comment:
Should this:

```csharp
var allowUnsafeBlock = propertyGroupNode.Element("AllowUnsafeBlock");
```

Actually be

var allowUnsafeBlock = propertyGroupNode.Element("AllowUnsafeBlock**s**");

Notice the "s" in "Blocks".

When I search for this I see "did you mean AllowUnsafeBlocks" instead:

https://www.google.com/search?q=AllowUnsafeBlock

Reply:
Yes, absolutely! Was a typo. A few lines down (L543) it is correct. That's what you get for once not copy&pasting 😅